lib/messages.coffee

```diff
@@ -1,121 +1,60 @@
-Helpers = require('./helpers')
-{CompositeDisposable, Emitter} = require('atom')
-
-###
-  Note: We are reclassifying the messages on on Pane Item Change,
-  even though we are relinting on that same event, 'cause linters
-  could take time and we have to refresh the views immediately
-###
+{Emitter, TextEditor} = require('atom')
+validate = require('./validate')
+helpers = require('./helpers')
 
 class MessageRegistry
-  constructor: (@linter) ->
-    @count = File: 0, Project: 0
-    @messages = new Map() # Messages = Map<Linter, Array<message>>
+  constructor: ->
+    @updated = false
+    @publicMessages = []
     @emitter = new Emitter
-    @subscriptions = new CompositeDisposable
-    @subscriptions.add @linter.observeEditorLinters (EditorLinter) =>
-      EditorLinter.onShouldUpdate ({linter, results: messages}) =>
-        Helpers.validateMessages(messages)
-        @classifyMessages(messages)
-        if EditorLinter.messages.has(linter)
-          @countMessages(EditorLinter.messages.get(linter), false)
-        EditorLinter.messages.set(linter, messages)
-        @countMessages(messages)
-        @emitter.emit 'did-change'
-      EditorLinter.onDidDestroy =>
-        EditorLinter.messages.forEach (messages) =>
-          @countMessages(messages, false)
-          @emitter.emit 'did-change'
-    @subscriptions.add atom.workspace.onDidChangeActivePaneItem =>
-      @count = File: 0, Project: 0
-      @messages.forEach (messages) =>
-        @classifyMessages(messages)
-        @countMessages(messages)
-      @linter.eachEditorLinter (EditorLinter) =>
-        EditorLinter.messages.forEach (messages) =>
-          @classifyMessages(messages)
-          @countMessages(messages)
-      @emitter.emit 'did-change'
+    @messages = new Map()
+    @editorMessages = new Map()
 
-  set: (linter, messages) ->
-    Helpers.validateMessages(messages)
-    @classifyMessages(messages)
-    if @messages.has(linter)
-      @countMessages(@messages.get(linter), false)
-    @messages.set(linter, messages)
-    @countMessages(messages)
-    @emitter.emit 'did-change'
+    @shouldUpdatePublic = true
+    requestAnimationFrame => @updatePublic()
 
-  delete: (linter) ->
+  set: ({linter, messages, editor}) ->
+    try validate.messages(messages) catch e then return helpers.error(e)
+    if linter.scope is 'project'
+      @messages.set(linter, messages)
+    else
+      throw new Error("Given editor isn't really an editor") unless editor instanceof TextEditor
+      if not @editorMessages.has(editor) then @editorMessages.set(editor, new Map())
+      @editorMessages.get(editor).set(linter, messages)
+    @updated = true
+
+  updatePublic: ->
+    return unless @shouldUpdatePublic
+    if @updated
+      @updated = false
+      publicMessages = []
+      @messages.forEach (messages) -> publicMessages = publicMessages.concat(messages)
+      @editorMessages.forEach (linters) -> linters.forEach (messages) ->
```

Comment:
Because keeping it this way helps us replace messages by the previous linter. `@editorMessages` has a shape of `array<TextEditor, array<Linter, array<Messages>>>`